### DISCUSSION TITLE
A question about CallTraceStorage::calcHash

### DISCUSSION THREAD

**Discussion by yanglong1010** [async-profiler/async-profiler#1089] [General]
Hi,

calcHash treats`ASGCT_CallFrame* frames` as u64 array, and iterates from start to end to calculate hash.
https://github.com/async-profiler/async-profiler/blob/8da7831bb1fbd5943ab1f5c5b54f35a18f2a7bf5/src/callTraceStorage.cpp#L175

If cpu and lock engines both are enabled, and for the same thread, is it possible the lock engine pollutes the CallTraceBuffer,  causes the cpu engine to get different hash for the same stack trace ? 

- time 0, cpu engine treats CallTraceBuffer as ASGCT_CallFrame array, for every frame, read first 4 bytes bci + 4 bytes padding as u64
- time 1, lock engine records sample, treats CallTraceBuffer as jvmtiFrameInfo array, write 8 bytes jmethodID to the first 8 bytes, key point, byte 4 to 7 is modified.
- time 2, cpu engine treats CallTraceBuffer as ASGCT_CallFrame array again, for every frame, read first 4 bytes bci + 4 bytes padding as u64, but now, though the first 4 bytes is the same, the other 4 bytes is different from time 0.

```c
union CallTraceBuffer {
    ASGCT_CallFrame _asgct_frames[1];
    jvmtiFrameInfo _jvmti_frames[1];
};

typedef struct {
    jint bci; // 4 bytes
    // 4 bytes padding;
    jmethodID method_id;
} ASGCT_CallFrame;

struct jvmtiFrameInfo {
    jmethodID method; // 8bytes
    jlocation location;
};
```

Thanks.

**Comment by apangin** [comment]
Nice observation! The scenario you've described is plausible. I haven't noticed such a problem, but it is difficult to detect, especially since it does not cause any visible effects, just inefficient use of CallTraceStorage.

I see 3 options to fix this:
1. Use different buffers for AGCT and JVMTI (requires additional memory).
2. Use only real `ASGCT_CallFrame` fields when computing hash (may complicate the algorithm or make it slower).
3. Add an explicit padding field on 64-bit platforms and fill it with zero when converting from JVMTI to AGCT frames (a bit of a hack, but probably most efficient).

I'd go for the 3rd option. Any thoughts?

**Reply by yanglong1010** [reply]
Yes, 3rd option is the most efficient, it only adds an assignment to the existing converting loop.

**Reply by yanglong1010** [reply]
I submitted a [PR](https://github.com/async-profiler/async-profiler/pull/1090) for this discussion.

**Comment by yanglong1010** [comment]
In addition, when I do the `cpu profiling` testing, byte 4 to 7 for every frame is always 0,  but I don't see any memory clearing code for the `_calltrace_buffer`. Did I miss some details? If the bytes happen to be zero, I think we need to zero the bytes explicitly.

```
(gdb) bt
#0  CallTraceStorage::calcHash (this=0x7fff70021308, num_frames=291, frames=0x7fff7080a640) at src/callTraceStorage.cpp:174
#1  0x00007fffd839235e in CallTraceStorage::put (this=0x7fff70021308, num_frames=291, frames=0x7fff7080a640, counter=10007039) at src/callTraceStorage.cpp:228
#2  0x00007fffd83adb34 in Profiler::recordSample (this=0x7fff700191f0, ucontext=0x7fffd89346c0, counter=10007039, event_type=PERF_SAMPLE, event=0x7fffd8934690) at src/profiler.cpp:706
#3  0x00007fffd83aa8b5 in PerfEvents::signalHandler (signo=27, siginfo=0x7fffd89347f0, ucontext=0x7fffd89346c0) at src/perfEvents_linux.cpp:688
#4  <signal handler called>
#5  0x00007ffff7ffacf7 in gettimeofday ()
#6  0x00007ffff6a596d2 in os::javaTimeMillis() () from /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.412.b08-1.el7_9.x86_64/jre/lib/amd64/server/libjvm.so
#7  0x00007fffe116fe10 in ?? ()
#8  0x000002cf000002d0 in ?? ()
#9  0x00000007c0001ef8 in ?? ()
#10 0x00000007c00578b0 in ?? ()

(gdb) x/64x data                 
0x7fff7080a640: 0xfffffff6      0x00000000      0x70758474      0x00007fff
0x7fff7080a650: 0xfffffff6      0x00000000      0x703afab4      0x00007fff
0x7fff7080a660: 0x05000023      0x00000000      0x7076a278      0x00007fff
0x7fff7080a670: 0x05000011      0x00000000      0x7076a298      0x00007fff
0x7fff7080a680: 0x03000032      0x00000000      0x70769fb8      0x00007fff
0x7fff7080a690: 0x00000043      0x00000000      0x70769fb8      0x00007fff
0x7fff7080a6a0: 0x00000043      0x00000000      0x70769fb8      0x00007fff
0x7fff7080a6b0: 0x00000043      0x00000000      0x70769fb8      0x00007fff
0x7fff7080a6c0: 0x00000043      0x00000000      0x70769fb8      0x00007fff
0x7fff7080a6d0: 0x00000043      0x00000000      0x70769fb8      0x00007fff
0x7fff7080a6e0: 0x00000043      0x00000000      0x70769fb8      0x00007fff
0x7fff7080a6f0: 0x00000043      0x00000000      0x70769fb8      0x00007fff
0x7fff7080a700: 0x00000043      0x00000000      0x70769fb8      0x00007fff
0x7fff7080a710: 0x00000043      0x00000000      0x70769fb8      0x00007fff
```

**Reply by apangin** [reply]
You're right, it's better to replace `malloc` with `calloc`:
https://github.com/async-profiler/async-profiler/blob/8da7831bb1fbd5943ab1f5c5b54f35a18f2a7bf5/src/profiler.cpp#L1137

**Comment by yanglong1010** [comment]
Thank you for your reply and review.
